File: app/src/main/java/com/nirbhor/app/ui/screens/family/CaregiverCodeScreen.kt
package com.nirbhor.app.ui.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.nirbhor.app.navigation.NavActions
import com.nirbhor.app.ui.components.PrimaryButton
import com.nirbhor.app.ui.components.TintPanel
import com.nirbhor.app.ui.components.NirbhorTopBar
import com.nirbhor.app.ui.i18n.tr
import com.nirbhor.app.ui.theme.Archivo
import com.nirbhor.app.ui.theme.Dimens
import com.nirbhor.app.ui.theme.NirbhorTheme

/** Caregiver joins by code (6d): four code tiles, then a confirmation + permission disclosure. */
@Composable
fun CaregiverCodeScreen(actions: NavActions) {
    val colors = NirbhorTheme.colors
    var code by remember { mutableStateOf("") }
    val matched = code.length == 4

    Column(Modifier.fillMaxSize().background(colors.paper)) {
        NirbhorTopBar(title = tr("কোড দিয়ে যুক্ত হোন", "Join by code"), onBack = actions::back)
        Column(
            Modifier.fillMaxSize().verticalScroll(rememberScrollState()).navigationBarsPadding()
                .padding(horizontal = Dimens.screenPadding, vertical = 20.dp),
            verticalArrangement = Arrangement.spacedBy(Dimens.groupGap), horizontalAlignment = Alignment.CenterHorizontally,
        ) {
            Text(tr("রোগীর অ্যাপে দেখানো ৪-অক্ষরের কোডটি লিখুন", "Enter the 4-character code shown in the patient's app"),
                style = NirbhorTheme.type.body, color = colors.ink2)

            Box {
                Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                    for (i in 0 until 4) {
                        val c = code.getOrNull(i)?.toString() ?: ""
                        val active = i == code.length
                        Box(
                            Modifier.size(70.dp).clip(RoundedCornerShape(14.dp)).background(colors.card)
                                .border(if (active) 2.dp else 1.5.dp, if (active) colors.calm else colors.line, RoundedCornerShape(14.dp)),
                            contentAlignment = Alignment.Center,
                        ) {
                            Text(c, fontSize = 28.sp, fontWeight = FontWeight(700), fontFamily = Archivo, color = colors.ink)
                        }
                    }
                }
                BasicTextField(
                    value = code,
                    onValueChange = { if (it.length <= 4) code = it.uppercase() },
                    keyboardOptions = KeyboardOptions(capitalization = KeyboardCapitalization.Characters),
                    modifier = Modifier.matchParentSize().alpha(0f),
                )
            }

            if (matched) {
                TintPanel(background = colors.calmSoft, radius = Dimens.radiusLargeCard) {
                    Text(tr("আব্বার সঙ্গে যুক্ত হচ্ছেন", "Connecting to Abba"), style = NirbhorTheme.type.cardTitlePrimary, color = colors.calmD)
                }
                TintPanel(background = colors.sage) {
                    listOf(
                        tr("• আপনি শুধু দেখতে পারবেন, বদলাতে পারবেন না", "• You can only view, not change anything"),
                        tr("• রোগী চাইলে যেকোনো সময় সরাতে পারবেন", "• The patient can remove you any time"),
                        tr("• আপনি যা দেখছেন, রোগীও তা জানেন", "• The patient sees what you see"),
                    ).forEach { Text(it, style = NirbhorTheme.type.body, color = colors.ink2, modifier = Modifier.padding(vertical = 2.dp)) }
                }
            }

            Spacer(Modifier.height(24.dp))
            PrimaryButton(tr("যুক্ত হোন", "Join"), actions::back, height = 68.dp, enabled = matched, modifier = Modifier.fillMaxWidth())
            Text(tr("কোড নেই", "No code"), style = NirbhorTheme.type.buttonLabel, color = colors.ink3, modifier = Modifier.padding(6.dp))
        }
    }
}

File: lib/ui/screens/caregiver_code_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import '../../navigation/nav_actions.dart';
import '../../theme/theme.dart';
import '../components/buttons.dart';
import '../components/scaffold.dart';
import '../components/surfaces.dart';

/// Join by code (6d). The four boxes are a styled overlay on one hidden field, so the platform
/// keyboard, selection and paste all behave normally.
class CaregiverCodeScreen extends StatefulWidget {
  const CaregiverCodeScreen({super.key});

  @override
  State<CaregiverCodeScreen> createState() => _CaregiverCodeScreenState();
}

class _CaregiverCodeScreenState extends State<CaregiverCodeScreen> {
  final _controller = TextEditingController();
  final _focus = FocusNode();
  bool _joinAttempted = false;
  bool _showCodeHelp = false;

  String get _code => _controller.text;
  bool get _matched => _code.length == 4;

  @override
  void dispose() {
    _controller.dispose();
    _focus.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final colors = context.colors;
    return Scaffold(
      backgroundColor: colors.paper,
      body: Column(
        children: [
          NirbhorTopBar(
            title: context.tr('কোড দিয়ে যুক্ত হোন', 'Join by code'),
            onBack: context.nav.back,
          ),
          Expanded(
            child: SingleChildScrollView(
              padding: const EdgeInsets.symmetric(
                horizontal: Dimens.screenPadding,
                vertical: 20,
              ),
              child: SafeArea(
                top: false,
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.stretch,
                  children: [
                    Text(
                      context.tr(
                        'রোগীর অ্যাপে দেখানো ৪-অক্ষরের কোডটি লিখুন',
                        "Enter the 4-character code shown in the patient's app",
                      ),
                      style: context.type.body.copyWith(color: colors.ink2),
                      textAlign: TextAlign.center,
                    ),
                    const SizedBox(height: Dimens.groupGap),
                    Center(
                      child: Stack(
                        children: [
                          Row(
                            mainAxisSize: MainAxisSize.min,
                            children: [
                              for (var i = 0; i < 4; i++) ...[
                                if (i > 0) const SizedBox(width: 10),
                                Container(
                                  width: 70,
                                  height: 70,
                                  alignment: Alignment.center,
                                  decoration: BoxDecoration(
                                    color: colors.card,
                                    borderRadius: BorderRadius.circular(14),
                                    border: Border.all(
                                      color: i == _code.length ? colors.calm : colors.line,
                                      width: i == _code.length ? 2 : 1.5,
                                    ),
                                  ),
                                  child: Text(
                                    i < _code.length ? _code[i] : '',
                                    // The code is Latin in both locales.
                                    style: context.type
                                        .asLatin(context.type.titleHero)
                                        .copyWith(fontSize: 28, color: colors.ink),
                                  ),
                                ),
                              ],
                            ],
                          ),
                          Positioned.fill(
                            child: Opacity(
                              opacity: 0,
                              child: TextField(
                                controller: _controller,
                                focusNode: _focus,
                                autofocus: true,
                                showCursor: false,
                                textCapitalization: TextCapitalization.characters,
                                inputFormatters: [
                                  LengthLimitingTextInputFormatter(4),
                                  TextInputFormatter.withFunction(
                                    (_, next) => next.copyWith(text: next.text.toUpperCase()),
                                  ),
                                ],
                                onChanged: (_) => setState(() => _joinAttempted = false),
                              ),
                            ),
                          ),
                        ],
                      ),
                    ),
                    if (_matched) ...[
                      const SizedBox(height: Dimens.groupGap),
                      TintPanel(
                        background: colors.calmSoft,
                        radius: Dimens.radiusLargeCard,
                        // Nothing has resolved this code yet, so name no one here.
                        child: Text(
                          context.tr(
                            'রোগীর হিসাবের সঙ্গে যুক্ত হচ্ছেন',
                            "Connecting to the patient's record",
                          ),
                          style: context.type.cardTitlePrimary.copyWith(color: colors.calmD),
                        ),
                      ),
                      const SizedBox(height: Dimens.cardGap),
                      TintPanel(
                        background: colors.sage,
                        child: Column(
                          crossAxisAlignment: CrossAxisAlignment.start,
                          children: [
                            for (final line in [
                              context.tr(
                                '• আপনি শুধু দেখতে পারবেন, বদলাতে পারবেন না',
                                '• You can only view, not change anything',
                              ),
                              context.tr(
                                '• রোগী চাইলে যেকোনো সময় সরাতে পারবেন',
                                '• The patient can remove you any time',
                              ),
                              context.tr(
                                '• আপনি যা দেখছেন, রোগীও তা জানেন',
                                '• The patient sees what you see',
                              ),
                            ])
                              Padding(
                                padding: const EdgeInsets.symmetric(vertical: 2),
                                child: Text(
                                  line,
                                  style: context.type.body.copyWith(color: colors.ink2),
                                ),
                              ),
                          ],
                        ),
                      ),
                    ],
                    const SizedBox(height: 24),
                    if (_joinAttempted) ...[
                      TintPanel(
                        background: colors.warmSoft,
                        child: Text(
                          context.tr(
                            'কোড দিয়ে যুক্ত হওয়া এখনও চালু হয়নি। এই সংস্করণে হিসাব শুধু রোগীর ফোনেই থাকে।',
                            "Joining by code isn't switched on yet. In this version the record stays on the patient's phone.",
                          ),
                          style: context.type.body.copyWith(color: colors.warmD),
                        ),
                      ),
                      const SizedBox(height: 12),
                    ],
                    PrimaryButton(
                      text: context.tr('যুক্ত হোন', 'Join'),
                      height: 68,
                      enabled: _matched,
                      onPressed: () => setState(() => _joinAttempted = true),
                    ),
                    TextButton(
                      onPressed: () => setState(() => _showCodeHelp = !_showCodeHelp),
                      style: TextButton.styleFrom(
                        minimumSize: const Size.fromHeight(Dimens.tapMin),
                      ),
                      child: Text(
                        context.tr('কোড নেই?', 'No code?'),
                        style: context.type.buttonLabel.copyWith(color: colors.calm),
                      ),
                    ),
                    if (_showCodeHelp)
                      TintPanel(
                        background: colors.sage,
                        child: Text(
                          context.tr(
                            'রোগীর ফোনে নির্ভর খুলে “আরও → পরিবার ও যত্নকারী”-তে যেতে বলুন। সেখানে ৪-অক্ষরের কোডটি দেখা যাবে।',
                            'Ask the patient to open Nirbhor and go to More → Family & caregivers. The 4-character code is shown there.',
                          ),
                          style: context.type.body.copyWith(color: colors.ink2),
                        ),
                      ),
                  ],
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}
